File: desktop/src/homemade/game/fieldstructure/CellCodePair.java
package homemade.game.fieldstructure;

public class CellCodePair {

    private CellCode lower;
    private CellCode higher;

    public CellCodePair(LinkCode link) {
        this.lower = link.getLower();
        this.higher = link.getHigher();
    }

    public void move(Direction direction) {
        lower = lower.neighbour(direction);
        higher = higher.neighbour(direction);
    }

    public boolean isValid() {
        return lower != null && higher != null;
    }

    public CellCode getLower() {
        return lower;
    }

    public CellCode getHigher() {
        return higher;
    }
}

File: field-structure/src/main/java/homemade/game/fieldstructure/CellCodePair.kt
package homemade.game.fieldstructure

class CellCodePair(link: LinkCode) {
    var lower: CellCode?
        private set
    var higher: CellCode?
        private set

    init {
        lower = link.lower
        higher = link.higher
    }

    fun move(direction: Direction) {
        if (isValid) {
            lower = lower!!.neighbour(direction)
            higher = higher!!.neighbour(direction)
        }
    }

    val isValid: Boolean
        get() = lower != null && higher != null
}
